client-report node and package and upgrades; eslint; dev server

ballPointPenguin opened the pull request:
Continuation of earlier work to upgrade client-report to LTS Node 18.
https://github.com/compdemocracy/polis/pull/1541
Updates to babel and eslint deprecations.
Get webpack-dev-server working.
Remove SCP stuff.
Take care of all eslint errors.
Some modest React 17 updates (PropTypes and ref="")

How I tested this:
To be fair I didn't build any reports. We should definitely check that workflow before merging.
I was able to run `npm install` and **all** of the npm scripts, including:
`npm run build:prod`
`npm run bundle`
`npm start`
`npm run lint`

I set react/prop-types lint rule to "warn" and made it so `npm run lint` will only show errors.
Future work: run `npm run lint:verbose` and fix all of the PropTypes.

Additionally I spot checked the docker build with:
`docker build -t polis-client-server:local .`
`docker run --rm polis-client-server:local`
which completed without issues.

dependabot[bot] committed:
Merge branch 'edge' into colinmegill/client-report-package-upgrades
* edge: (22 commits)
  Configuration unification (#1617)
  housekeeping (#1622)
  Remove intercom (#1547)
  Reduced bundlesize by requiring only victory-core for VictoryAnimation component. (#1618)
  A variety of improvements all around (#1608)
  Update client-participation dependencies, and get off gulp -> webpack (#1607)
  Feature/crkrenn/fixed participants extended sql table (#1586)
  Switch server to debian (#1600)
  More slack removal (#1599)
  Update server node & misc dep versions (#1595)
  switch from npm ci to npm install in client dockerfiles (#1594)
  deleted docker_cloud_notes (#1597)
  Remove slack (#1585)
  specify bullseye debian base for math dockerfile (#1590)
  Feature/crkrenn/improved makefile (#1577)
  Adding direction to the client- participation (#1561)
  update winston package (#1584)
  add scaling roadmap doc (#1582)
  Feature/sirodoht crk/e2e version 11 (#1576)
  German language editing (#1552)
  ...

dependabot[bot] committed:
Bump moment from 2.29.1 to 2.29.4 in /client-report (#1514)

Bumps [moment](https://github.com/moment/moment) from 2.29.1 to 2.29.4.
- [Release notes](https://github.com/moment/moment/releases)
- [Changelog](https://github.com/moment/moment/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/moment/moment/compare/2.29.1...2.29.4)

---
updated-dependencies:
- dependency-name: moment
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump ws from 7.4.2 to 7.5.7 in /client-admin (#1395)

Bumps [ws](https://github.com/websockets/ws) from 7.4.2 to 7.5.7.
- [Release notes](https://github.com/websockets/ws/releases)
- [Commits](https://github.com/websockets/ws/compare/7.4.2...7.5.7)

---
updated-dependencies:
- dependency-name: ws
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump got and nodemon in /server (#1539)

Removes [got](https://github.com/sindresorhus/got). It's no longer used after updating ancestor dependency [nodemon](https://github.com/remy/nodemon). These dependencies need to be updated together.

Removes `got`

Updates `nodemon` from 2.0.7 to 2.0.20
- [Release notes](https://github.com/remy/nodemon/releases)
- [Commits](https://github.com/remy/nodemon/compare/v2.0.7...v2.0.20)

---
updated-dependencies:
- dependency-name: got
  dependency-type: indirect
- dependency-name: nodemon
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump minimist from 1.2.5 to 1.2.6 in /client-admin (#1384)

Bumps [minimist](https://github.com/substack/minimist) from 1.2.5 to 1.2.6.
- [Release notes](https://github.com/substack/minimist/releases)
- [Commits](https://github.com/substack/minimist/compare/1.2.5...1.2.6)

---
updated-dependencies:
- dependency-name: minimist
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

colinmegill committed:
babelrc update

colinmegill committed:
prettier

colinmegill committed:
prettier 2

colinmegill committed:
remove scp2

colinmegill committed:
prettier dev server

colinmegill committed:
update all packages and webpack accordingly

ballPointPenguin committed:
pretend that scp never existed

ballPointPenguin committed:
remove unused X-File

ballPointPenguin committed:
proper html5 meta

ballPointPenguin committed:
further update packages, using tilde semver;
replace new @babel packages;
touch up the npm scripts;
remove postinstall script;
include graceful-fs for node+gulp compatibility.

ballPointPenguin committed:
freshen up the eslint config

ballPointPenguin committed:
get the webpack-dev-server working

ballPointPenguin committed:
ignore the "devel" directory

ballPointPenguin committed:
eslint all the things

ballPointPenguin committed:
simply use node:18-alpine for Dockerfile

ballPointPenguin committed:
update file-server Dockerfile node version

ballPointPenguin committed:
remove unused client-report/bin files

ballPointPenguin committed:
rebuild package-lock.json

ballPointPenguin committed:
remove comment about gulp in Dockerfile

ballPointPenguin committed:
Merge branch 'edge' into br/client-report-package-upgrades

ballPointPenguin committed:
Merge branch 'edge' into br/client-report-package-upgrades

* edge:
  toggle 3rd party services (#1639)

ballPointPenguin committed:
Merge branch 'edge' into br/client-report-package-upgrades
* edge:
  1379 clean up server logging (#1629)
  Clojure dep updates (#1625)

metasoarous commented:
Still getting CORS errors when I build/run, which is where I got hung up last time I looked at this PR

![image](https://user-images.githubusercontent.com/88556/222882177-e260131f-1c92-410c-b141-38ceffcd2b35.png)

ballPointPenguin committed:
update Color syntax in client-report

ballPointPenguin committed:
clean up url.js and do a more intelligent domain determination;
allow reports to be hosted at localhost with no port;
remove hard-coded urls from client-report;
rename REPORT_SERVICE_URL back to SERVICE_URL;
remove localhost:5002 references;

ballPointPenguin committed:
Merge branch 'edge' into br/client-report-package-upgrades
* edge:
  fool me twice... (#1654)
  ensure response is returned when authenticating user (#1652)
  check before firing off these Twitter/FB requests (#1651)
  replace _.zipObject with _.fromPairs (#1650)
  Remove Stripe/Slack/Canvas from codebase and db (#1620)
  comment out server from math/system; hotfix
  rm rm-ALL-ALL-TAGS; it's extremely dangerous (#1649)
  hot fix for math worker

ballPointPenguin committed:
include forward slash after serviceUrl

ballPointPenguin committed:
copypasta repair (#1655)

metasoarous commented on client-report/Dockerfile:
```diff
@@ -23,6 +22,14 @@ COPY . .
 # GIT_HASH is optional and will be set properly when running `make` (see Makefile).
 # Or may be passed in at build time with --build-arg GIT_HASH=$(git rev-parse --short HEAD)
 ARG GIT_HASH
+ARG REPORT_UPLOADER
+ARG S3_BUCKET_PREPROD
+ARG S3_BUCKET_PROD
```
These three can go; None of that is happening here any more.

metasoarous replied:
Just need to keep `SERVICE_URL` (and maybe `GIT_HASH`)

ballPointPenguin committed:
fix URL rendering in report heading

ballPointPenguin committed:
get rid of s3 buckets and the concept of 'uploader';
simplify the gulpfile;
update the README

ballPointPenguin committed:
Merge branch 'edge' into br/client-report-package-upgrades

metasoarous commented:
I see that we still have a `client-report/dev-server.js`. In the other webpack conversions, we've gone for just leveraging the built in dev server. Should probably do that here as well to further remove unecessary code.

metasoarous commented on client-report/Dockerfile:
```diff
@@ -23,6 +22,8 @@ COPY . .
+ENV SERVICE_URL ${SERVICE_URL}
 
-CMD npm run deploy:prod
+CMD npm run build:prod
```
Need to circle back around and make sure that cache files are getting tagged for cache busting; This was one thing that happened in `deploy:prod` but not `build:prod` on some of the clients.

metasoarous commented on client-report/index.html:
```diff
@@ -1,87 +1,89 @@
+  </div>
+  <script src="https://d3js.org/d3.v4.min.js"></script>
+  <script src="https://d3js.org/d3-scale-chromatic.v1.min.js"></script>
+  <script src="/dist/report_bundle.js"></script>
```
Regarding last comment, seems we already weren't doing the cache busting here, so maybe we're fine on that front, but possibly worth flagging whether we want to start doing that, and have webpack inject the script tag, as with participation (and I think admin) client(s). Traditionally, we've assumed that the report will be viewed by fewer people, having less load, and therefore less need for caching. But we're seeing an increasing trend of deployments that may be assuming open public access to reports, and should revisit these assumptions.

cc @xeeg

metasoarous replied:
Erm... nevermind. Looks like we were doing cache busting, so probably want to set up webpack as in participation and admin.

https://github.com/compdemocracy/polis/pull/1644/files#r1131615626

metasoarous commented on client-report/gulpfile.js:
```diff
@@ -1,430 +1,223 @@
-});
-
-
-gulp.task("scripts", function() {
-  var files = ["dist/report_bundle.js"];
-  var s = gulp.src(files);
-  s = s
-    .pipe(gzip())
-    .pipe(rename(function (path) {
-      // remove .gz extension
-      var ext = path.extname;
-      console.log("foo", path);
-      path.extname = ext.substr(0, ext.length- ".gz".length);
-    }));
-  return s.pipe(gulp.dest(destRoot() + "/js"));
-});
```
Actually looks like we were doing the cache busting tags here...

metasoarous commented on client-report/package.json:
```diff
@@ -4,83 +4,61 @@
-    "deploy:prod": "gulp deploy_TO_PRODUCTION",
+    "bundle:prod": "webpack --mode=production",
+    "bundle": "npm run clean && npm run build:prod",
+    "build:prod": "gulp build",
```
Let's completely nuke gulp!

ballPointPenguin replied:
This will be covered when https://github.com/compdemocracy/polis/issues/90 is done.

Also worth looking at `polis-whitelabel` solution
https://mathdem.slack.com/archives/C0374J43QMT/p1679306674336269?thread_ts=1678786732.540639&cid=C0374J43QMT

metasoarous commented on client-report/src/components/app.js:
```diff
@@ -10,17 +10,17 @@ import _ from "lodash";
+// import CommentsGraph from "./commentsGraph/commentsGraph";
 import ParticipantsGraph from "./participantsGraph/participantsGraph";
-import BoxPlot from "./boxPlot/boxPlot";
+// import BoxPlot from "./boxPlot/boxPlot";
```
I know we're not using correlation matrix or comments graph anymore, but are we not still using boxPlot for the report?

metasoarous replied:
Looked, and indeed does not seem used.

metasoarous commented on client-report/src/components/app.js:
```diff
@@ -136,8 +136,7 @@ class App extends React.Component {
           }, this.corMatRetries < 10 ? 200 : 3000); // try to get a quick response, but don't keep polling at that rate for more than 10 seconds.
         } else if (globals.enableMatrix && response && response.status === "polis_report_needs_comment_selection") {
           this.setState({
-            errorLink: "http://localhost:5002/m/36jajfnhhn/reports/r6ehukhk29tcfmuc57vrj/comments",
```
Good catch; Wonder how that got hardcoded in there.

xeeg commented:
Probably should update the "Quick Start" section of the main README.md from `make start` to `make start-FULL-REBUILD`?

ballPointPenguin committed:
Comment out problematic math command so it can run
ignore math errorconv files

ballPointPenguin commented:
> Still getting CORS errors when I build/run, which is where I got hung up last time I looked at this PR
> 
> ![image](https://user-images.githubusercontent.com/88556/222882177-e260131f-1c92-410c-b141-38ceffcd2b35.png)

FYI This is resolved.

ballPointPenguin commented:
> Probably should update the "Quick Start" section of the main README.md from `make start` to `make start-FULL-REBUILD`?

My thought is that for a "Quick Start" they won't have any stale volumes/containers to remove so won't need to FULL REBUILD. I hope that most people in most cases don't need to use FULL REBUILD since, in theory, they are building mainly  from one branch and mainly with source code that is not changing dramatically. FULL REBUILD is especially for us, who are frequently switching branches and rebuilding our containers to test significant changes.

xeeg commented on example.env:
Changing `MATH_ENV=dev` to `MATH_ENV=prod` here fixes the report generation problem. Due to some legacy conditions the math engine used to tag report records with prod or preprod and uses MATH_ENV to read them but now they're hardcoded to prod, however we have some records in production with preprod so we need to open another issue to properly migrate the data and remove the column.

ballPointPenguin replied:
OK. For now I changed the default MATH_EN to `prod`.

Testing with:

```sh
cp example.env .env
make start-FULL-REBUILD
```

and 

```sh
cp example.env prod.env
make PROD start-FULL-REBUILD
```

ballPointPenguin committed:
change default MATH_ENV to "prod"
